src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,472 @@
+in the directory that contains the package's `pubspec.yaml` file:
+
+```terminal
+$ dart migrate
+```
+
+If your package is ready to migrate,
+then the tool produces a line like the following:
+
+```terminal
+View the migration suggestions by visiting:
+
+  http://127.0.0.1:60278/Users/you/project/mypkg.console-simple?authToken=Xfz0jvpyeMI%3D
+```
+
+Visit that URL in a Chrome browser
+to see an interactive UI
+where you can guide the migration process:
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+
+For every variable and type annotation,
+you can see what nullability the tool infers.
+For example, in the preceding screenshot,
+the tool infers that the `ints` list (previously a list of `int`)
+in line 1 is nullable, and thus should be a list of `int?`.
+
+
+#### Understanding migration results
+
+To see the reasons for each change (or non-change),
+click its line number in the **Proposed Edits** pane.
+The reasons appear in the **Edit Details** pane.
+
+For example, consider the following code,
+from before null safety:
+
+```dart
+var ints = const <int>[0, null];
+var zero = ints[0];
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+
+The default migration is backward compatible but not ideal:
+
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+
+By clicking line 3, you can see the migration tool's reasons for
+adding the `!`.
+But if you know that `zero` can't be null,
+then you can improve the migration result.
+
+
+#### Improving migration results {#hint-markers}
+
+When analysis infers the wrong nullability,
+you can override its proposed edits by inserting temporary hint markers:
+
+* In the **Edit Details** pane of the migration tool,
+  you can insert hint markers using the
+  **Add `/*?*/` hint** and **Add `/*!*/` hint** buttons.
+
+  These buttons add comments to your file immediately,
+  and there's **no Undo**.
+  If you don't want a hint that the tool inserted,
+  you can use your usual code editor to remove it.
+
+* You can use an editor to add hint markers,
+  even while the tool is still running.
+  Because your code hasn't opted into null safety yet,
+  you can't use new null-safety features.
+  You can, however, make changes like refactoring
+  that don't depend on null-safety features.
+
+  When you've finished editing your code,
+  click **Rerun from sources** to pick up your changes.
+
+The following table shows the hint markers that you can use
+to change the migration tool's proposed edits.
+
+|------------------+--------------------------------------------------------------------------|
+| Hint marker      | Effect on the migration tool                                             |
+|------------------|--------------------------------------------------------------------------|
+| `/*!*/`          | Mark the preceding expression or type as non-null or non-nullable.       |
```

Comment:
This is a little misleading because the meaning of `/*!*/` is quite different depending whether you're applying it to an expression or a type.  If you're applying it to a type it means "don't make this type nullable", which I think is consistent with what you're saying here.  If you're applying it to an expression, it means "add a `!` here when migrating, so that the expression is null checked"; that's really different from marking the expression as non-nullable; in point of fact usually the only reason you would want to add a `!` to an expression is if you expect that the expression's type to be nullable (otherwise there would be no point in null checking it).

I wonder if it would be clearer to break these two cases out into separate entries in the table, e.g.:

| Hint marker | Effect on the migration tool |
|-|-|
| _expression_ `/*!*/` | Ensure that _expression_ is null checked in the migrated output. |
| _type_ `/*!*/` | Mark _type_ as non-nullable. |

Reply:
Thanks, @stereotype441!

Reply:
For the first one, we use different terminology in the null safety docs, so I'm thinking of something like this:

> Adds a `!` to the migrated code, which casts _expression_ to its underlying non-nullable type.

Reply:
> For the first one, we use different terminology in the null safety docs, so I'm thinking of something like this:
> 
> > Adds a `!` to the migrated code, which casts _expression_ to its underlying non-nullable type.

Excellent, that's way better than my suggestion 😃